### PR TITLE
refactor: use IterateLastValidatorPowers instead of GetLastValidators

### PR DESCRIPTION
**Pull request by p-offtermatt**
## Description

Closes: #1947 

This replaces the call to stakingKeeper.GetLastValidators with calls to providerKeeper.GetLastBondedValidators, which in turn is a wrapper around stakingKeeper.IterateLastValidatorPowers

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if the change is [state-machine breaking](https://github.com/cosmos/interchain-security/blob/main/RELEASES.md#breaking-changes)
* [x] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] Provided a link to the relevant issue or specification
* [x] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
No changelog needed since this is just a refactor?

* [x] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [x] Updated the relevant documentation or specification
* [x] Reviewed "Files changed" and left comments if necessary 
* [x] Confirmed all CI checks have passed
* [x] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` the type prefix if the change is state-machine breaking
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

## Summary by CodeRabbit

- **New Features**
  - Introduced new functions to retrieve and process bonded validators for improved consistency and efficiency.

- **Bug Fixes**
  - Updated various function calls to ensure correct retrieval of bonded validators, enhancing the reliability of validator-related operations.

- **Tests**
  - Enhanced test setups to provide more control and flexibility in testing scenarios involving bonded validators.
  - Added new parameters and updated logic in test functions to improve mock setups and expectations.

- **Refactor**
  - Replaced deprecated methods with new implementations across multiple files for better maintainability and performance.

### CODE REVIEW

File: testutil/keeper/expectations.go [hunk 1]
```diff
@@ -217,3 +217,36 @@ func GetMocksForSlashValidator(
 			Times(1),
 	}
 }
+
+// SetupMocksForLastBondedValidatorsExpectation sets up the expectation for the `IterateLastValidatorPowers` `MaxValidators`, and `GetValidator` methods of the `mockStakingKeeper` object.
+// These are needed in particular when calling `GetLastBondedValidators` from the provider keeper.
+// Times is the number of times the expectation should be called. Provide -1 for `AnyTimes“.
+func SetupMocksForLastBondedValidatorsExpectation(mockStakingKeeper *MockStakingKeeper, maxValidators uint32, vals []stakingtypes.Validator, powers []int64, times int) {
```
Comment (p-offtermatt) [inline testutil/keeper/expectations.go:224]: Open to suggestions for avoiding/improving the times parameter. I think it's not too bad, but not very pretty code, either

File: x/ccv/provider/keeper/keeper.go [hunk 2]
```diff
@@ -1513,3 +1514,36 @@ func (k Keeper) IsDenylistEmpty(ctx sdk.Context, chainID string) bool {
...
+		val, found := k.stakingKeeper.GetValidator(ctx, addr)
+		if !found {
+			continue
+		}
+
+		// gather all the bonded validators in order to construct the consumer validator set for consumer chain `chainID`
+		bondedValidators = append(bondedValidators, val)
+	}
+	return bondedValidators
+}
```
[the remark this answers is not in the document]
Comment (p-offtermatt) [reply x/ccv/provider/keeper/keeper.go]: Good review

File: x/ccv/provider/keeper/proposal_test.go [hunk 3]
```diff
@@ -1199,7 +1201,7 @@
 	expectations := []*gomock.Call{}
 	for _, prop := range pendingProps {
 		// A consumer chain is setup corresponding to each prop, making these mocks necessary
-		mocks.MockStakingKeeper.EXPECT().GetLastValidators(gomock.Any()).Times(1)
+		testkeeper.SetupMocksForLastBondedValidatorsExpectation(mocks.MockStakingKeeper, 0, []stakingtypes.Validator{}, []int64{}, 1)
```
[the remark this answers is not in the document]
Comment (p-offtermatt) [reply x/ccv/provider/keeper/proposal_test.go:1204]: bad comment, this is not the times parameter but the maxValidators parameter.

File: x/ccv/provider/keeper/keeper.go [hunk 4]
```diff
@@ -1513,3 +1514,38 @@ func (k Keeper) IsDenylistEmpty(ctx sdk.Context, chainID string) bool {
 
 	return !iterator.Valid()
 }
+
+// GetLastBondedValidators iterates the last validator powers in the staking module
+// and returns the first MaxValidators many validators with the largest powers.
+func (k Keeper) GetLastBondedValidators(ctx sdk.Context) []stakingtypes.Validator {
+	var lastPowers []stakingtypes.LastValidatorPower
```
Comment (sainoe) [inline x/ccv/provider/keeper/keeper.go:1521]: nit: `lastPowers` can be initiated with a length set to `maxVals`.

File: x/ccv/provider/keeper/keeper.go [hunk 5]
```diff
@@ -1513,3 +1514,38 @@ func (k Keeper) IsDenylistEmpty(ctx sdk.Context, chainID string) bool {
...
+	i := 0
+	k.stakingKeeper.IterateLastValidatorPowers(ctx, func(addr sdk.ValAddress, power int64) (stop bool) {
+		lastPowers = append(lastPowers, stakingtypes.LastValidatorPower{Address: addr.String(), Power: power})
+		i++
+		return i >= int(maxVals)
+	})
+
+	var bondedValidators []stakingtypes.Validator
+
+	for _, p := range lastPowers {
```
Comment (sainoe) [inline x/ccv/provider/keeper/keeper.go:1534]: Why do we need a second loop?

Comment (p-offtermatt) [reply x/ccv/provider/keeper/keeper.go:1534]: I think it's preferable to have as little code as possible inside the Iterate loop, but maybe you disagree.

Concretely, the first loop just iterates over powers+addresses, and this now gets the validator objects from store

File: x/ccv/provider/keeper/keeper.go (hunk 4 above)
Comment (p-offtermatt) [reply x/ccv/provider/keeper/keeper.go:1521]: Sure, can change

File: x/ccv/provider/keeper/keeper.go (hunk 5 above)
Comment (sainoe) [reply x/ccv/provider/keeper/keeper.go]: Okay, that's fine with me! We can never be too careful with these iterators :)

**Comment by sainoe** [comment]
very nit: `GetLastValidators` can be safely removed in the `StakingKeeper` interface definition in `expected_keeper.go` :)

LGTM. Thanks for the work!

**Review by sainoe — APPROVED** [review]

File: x/ccv/provider/keeper/keeper.go [hunk 6]
```diff
@@ -1513,3 +1514,41 @@ func (k Keeper) IsDenylistEmpty(ctx sdk.Context, chainID string) bool {
...
+		val, found := k.stakingKeeper.GetValidator(ctx, addr)
+		if !found {
+			k.Logger(ctx).Error("Validator not found", "address", addr.String())
+			continue
+		}
+
+		// gather all the bonded validators in order to construct the consumer validator set for consumer chain `chainID`
+		bondedValidators[index] = val
+	}
+	return bondedValidators
```
[the remark this answers is not in the document]
Comment (p-offtermatt) [reply x/ccv/provider/keeper/keeper.go]: Should we make a followup PR for this? I don't think we should solve this all in this PR, lmk if you disagree

File: x/ccv/provider/keeper/validator_set_update.go [hunk 7]
```diff
@@ -189,3 +190,9 @@ func (k Keeper) FilterValidators(
 
 	return nextValidators
 }
+
+// GetLastBondedValidators iterates the last validator powers in the staking module
+// and returns the first MaxValidators many validators with the largest powers.
+func (k Keeper) GetLastBondedValidators(ctx sdk.Context) []stakingtypes.Validator {
```
Comment (mpoke) [inline x/ccv/provider/keeper/validator_set_update.go:196]: Why do we need this function? Can't we just call GetLastBondedValidatorsUtil directly?

File: x/ccv/consumer/keeper/keeper.go [hunk 8]
```diff
@@ -690,3 +690,9 @@ func (k Keeper) IsPrevStandaloneChain(ctx sdk.Context) bool {
 	store := ctx.KVStore(k.storeKey)
 	return store.Has(types.PrevStandaloneChainKey())
 }
+
+// GetLastBondedValidators iterates the last validator powers in the staking module
+// and returns the first MaxValidators many validators with the largest powers.
+func (k Keeper) GetLastBondedValidators(ctx sdk.Context) []stakingtypes.Validator {
```
Comment (mpoke) [inline x/ccv/consumer/keeper/keeper.go:696]: ditto: Why do we need this method?

**Review by mpoke — APPROVED** [review]

File: x/ccv/provider/keeper/validator_set_update.go (hunk 7 above)
Comment (p-offtermatt) [reply x/ccv/provider/keeper/validator_set_update.go:196]: I think the syntax of passing in the staking keeper is unusual, and this way is much, much closer to how one would functions like this expect to look like. It's not necessary, but I think the interface is nicer like this.

**Pull request merged**